translation_layer/jpl-capi.js

```diff
@@ -0,0 +1,150 @@
+/*globals console, window*/
+define ('main',['require','jquery','ExtendedModel','api/snapshot/adapters/BackboneAdapter','sim-common/Detect','sim-common/VersionCheck','text!sim-common/templates/notSupported.html','api/snapshot/adapters/BackboneAdapter','api/snapshot/Controller','api/snapshot/Transporter','api/snapshot/SimCapiValue'],function(require){
+
+    var $ = require('jquery');
+    var $body;
+
+    var ExtendedModel = require('ExtendedModel');
+    var adapter = require('api/snapshot/adapters/BackboneAdapter').getInstance();
+
+    var Capi = ExtendedModel.extend({});
+    var capi;
+
+    require('sim-common/Detect');
+    require('sim-common/VersionCheck');
+
+    var notSupportedTemplate = require('text!sim-common/templates/notSupported.html');
+    var BackboneAdapter = require ('api/snapshot/adapters/BackboneAdapter').getInstance();
+    var Controller = require('api/snapshot/Controller');
+    var Transporter = require('api/snapshot/Transporter').getInstance();
+    var Types = require('api/snapshot/SimCapiValue').TYPES;
+
+	var initialized = false;
+
+    window.receiveExposeFromRanger = function (name, type, value, allowedValues)
+    {
+        if (!initialized)
+        {
+            Controller.notifyOnReady();
+            capi = new Capi();
+            capi.on('change', onCapiChange);
+            initialized = true;
+        }
+
+        var typeConverter = generateTypeConverter(type);
+        value = typeConverter(value);
+
+        var descObj = {};
+
+        //console.log("allowed values: " + allowedValues);
+
+        if (allowedValues)
+        {
+            descObj.allowedValues = allowedValues.map(typeConverter);
+        }
+
+        capi.set(name, value);
+        adapter.expose(name, capi, descObj);
+    };
+
+	    window.receiveValueFromRanger = function (name, type, value) {
+        if (initialized) {
+            value = generateTypeConverter(type)(value);
+            capi.set(name, value);
+        }
```

Comment:
I'm not sure exactly what you mean, and it's making me realize I'm not sure exactly how all this code is supposed to get called. As far as I see recieveValueFromRanger isn't used anywhere yet, so this comment doesn't matter at all! I suspect you expect it to be used though, but I'm not sure how. My initial comment was made because as far as I can tell, if this function is called before recieveExposeFromRanger the value would never be exposed.

You can make expose calls before notifyOnReady() is called, and nothing will get set in your CapiModel before notifyOnReady() is called, the viewer doesn't send anything to the sim before notifyOnReady(). Most of our sims do this in the opposite order though (so that we do the initial sim setup with the initial state values instead of default values) - we run notifyOnReady(), wait for the initialSetupComplete listener to be called, and then expose.

Additionally, I wasn't completely comfortable with this before but didn't mention it: receiveExposeFromRanger() feels overloaded - it is doing both the capi initialization and the exposes. It seems like you'd want to call notifyOnReady as soon as the page is ready to run javascript and then expose as Ranger initializes and sends you values.